site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md

```diff
@@ -0,0 +1,548 @@
+  }
+
+  updateNameTo = (value) => this.myName = value;
+}
+
+let personInfo = new PersonInfo("Chris");
+console.log(personInfo.nameLength); // 5
+
+personInfo.updateNameTo("Chris Krycho");
+console.log(personInfo.nameLength); // 12
+```
+
+But calling `personInfo.nameLength()` like this looks awfully familiar: it’s the same as the class method version we might have used before we had native getters. We’re back to where we started, in other words.
+
+The values a function uses are only evaluated when the function is invoked, whether the function in question is a standalone function, a class method, or a getter. If we have a *chain* of getters (or methods or functions), none of them will be reinvoked until the one at the end of the chain is. We won’t evaluate any of the values they reference until we access a getter which uses them. As a result, any time we evaluate a getter, we’ll always get an up-to-date version of all the values involved. We can add some logging to the getters in `PersonInfo` to see how this behaves:
+
+```js
+const MAX_LENGTH = 10;
+
+export default class PersonInfo {
+  name;
+
+  constructor(name) {
+    this.name = name;
+  }
+
+  nameLength() {
+    console.log("evaluating `nameLength`");
+    return this.name.length;
+  }
+
+  remaining() {
+    console.log("evaluating `remaining`");
+    return this.nameLength - MAX_LENGTH;
+  }
+
+  showError() {
+    console.log("evaluating `showError`");
+    return this.remaining < 0;
+  }
+
+  updateNameTo = (value) => this.myName = value;
+}
+```
+
+If we create and use a `PersonInfo` like this—
+
+```js
+let personInfo = new PersonInfo("Chris");
+console.log(" --- 1 --- ");
+console.log(personInfo.showError);
+
+console.log("\n --- 2 --- ");
+console.log(personInfo.nameLength);
+
+console.log("\n --- 3 --- ");
+personInfo.updateNameTo("Chris Krycho");
+console.log(personInfo.remaining);
+console.log(personInfo.showError);
+```
+
+—the console output would read:
+
+```
+" --- 1 --- "
+evaluating `nameLength`
+evaluating `remaining`
+evaluating `showError`
+false
+
+" --- 2 --- "
+evaluating `nameLength`
+5
+
+" --- 3 --- "
+evaluating `nameLength`
+evaluating `remaining`
+-2
+evaluating `nameLength`
+evaluating `remaining`
+evaluating `showError`
+true
+```
+
+In this example, the JavaScript I’ve written evaluates the values directly when logging them. When we use a value in a template in Ember or Glimmer apps, the template engine (the Glimmer VM) evaluates those values. The VM uses a lightweight reactivity system called *autotracking* to track which items in the UI need to be updated in any render. The next step, then, is understanding autotracking.
+
+[^hooks]: This is actually a critical part of how React Hooks work under the hood.
+
+[^closures-classes]: It’s also worth seeing how closures are the [dual](https://en.wikipedia.org/wiki/Duality_(mathematics)) of classes. These two have *the same semantics* as far as an end user is concerned:
+
+    ```js
```

Comment:
This example doesn't appear to highlight correctly

Reply:
Yeah, it’s highlightjs/highlight.js#2700. I’m debating on the trade off with just removing the private field in favor of a traditional “private” field, vs. leaving it as is and getting syntax highlighting once the upstream issue is fixed. Thoughts?

Reply:
I went with just using `_` names instead; the point is conveyed just as well that way and I've already clarified that this is not the real implementation. Better to just avoid distracting readers and move on.